library/ssl_tls13_generic.c

```diff
@@ -3940,6 +3971,53 @@ static int ssl_finished_out_postprocess( mbedtls_ssl_context* ssl )
             mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_CLIENT_CERTIFICATE );
         }
 
+        /* Compute hash over transcript of all messages sent up to the Finished message 
+         * sent by the server and store it in the digest variable of the handshake state. 
+         * This digest will be needed later when computing the application traffic secrets. 
+         */
+        cipher_info = mbedtls_cipher_info_from_type(ssl->transform_negotiate->ciphersuite_info->cipher );
+        if( cipher_info == NULL )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "cipher info for %d not found",
+                                        ssl->transform_negotiate->ciphersuite_info->cipher ) );
+            return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+        }
+
+        suite_info = mbedtls_ssl_ciphersuite_from_id( ssl->session_negotiate->ciphersuite );
+        if( suite_info == NULL )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "mbedtls_ssl_ciphersuite_from_id in mbedtls_ssl_derive_traffic_keys failed" ) );
+            return( MBEDTLS_ERR_SSL_BAD_HS_SERVER_HELLO );
+        }
+
+        if( mbedtls_hash_size_for_ciphersuite( suite_info ) == 32 )
+        {
+#if defined(MBEDTLS_SHA256_C)
+            mbedtls_sha256_init( &sha256 );
```

Comment:
`mbedtls_sha256_starts()` and `mbedtls_sha256_finish()` are deprecated; we should use `mbedtls_sha256_starts_ret()` and `mbedtls_sha256_finish_ret()` and check the return code. Example in `ssl_tls.c`:
```
    if( ( ret = mbedtls_md_setup( &ctx, md_info, 0 ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_md_setup", ret );
        goto exit;
    }
    if( ( ret = mbedtls_md_starts( &ctx ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_md_starts", ret );
        goto exit;
    }
    if( ( ret = mbedtls_md_update( &ctx, ssl->handshake->randbytes, 64 ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_md_update", ret );
        goto exit;
    }
    if( ( ret = mbedtls_md_update( &ctx, data, data_len ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_md_update", ret );
        goto exit;
    }
    if( ( ret = mbedtls_md_finish( &ctx, hash ) ) != 0 )
    {
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_md_finish", ret );
        goto exit;
    }```
Independently, a more generic API is used there. Should we use it too?

Reply:
I think we should do this as a separate PR because code needs to be changed in other places too.